File: workbook/ch07/data/rabin.py
def calculate_hash(s, length):
    hash_value = 0
    for i in range(length):
        hash_value = (hash_value * 256 + ord(s[i])) % 101  # prime 101
    return hash_value

def recalculate_hash(s, old_index, new_index, old_hash, pattern_length):
    new_hash = (old_hash - ord(s[old_index]) * (256 ** (pattern_length - 1))) % 101
    new_hash = (new_hash * 256 + ord(s[new_index])) % 101
    return new_hash if new_hash >= 0 else new_hash + 101

def rabin_karp_search(text, pattern):
    text_length = len(text)
    pattern_length = len(pattern)

    pattern_hash = calculate_hash(pattern, pattern_length)
    text_hash = calculate_hash(text, pattern_length)

    for i in range(text_length - pattern_length + 1):
        if pattern_hash == text_hash:
            if text[i:i + pattern_length] == pattern:
                print(f"Pattern found at index {i}")

        if i < text_length - pattern_length:
            text_hash = recalculate_hash(text, i, i + pattern_length, text_hash, pattern_length)


text = "ABABDABACDABABCABAB"
pattern = "ABABCABAB"

print(f"Text: {text}")
print(f"Pattern: {pattern}")

rabin_karp_search(text, pattern)

File: workbook/ch07/data/common/rabin.c
#include <stdio.h>
#include <string.h>

#define PRIME 101

long power(int base, int exponent) {
    long result = 1;
    for (int i = 0; i < exponent; i++) {
        result = (result * base) % PRIME;
    }
    return result;
}

long calculateHash(char *str, int length) {
    long hash = 0;
    for (int i = 0; i < length; i++) {
        hash = (hash * 256 + str[i]) % PRIME;
    }
    return hash;
}

long recalculateHash(char *str, int oldIndex, int newIndex, long oldHash, int patternLength) {
    long newHash = (oldHash - str[oldIndex] * power(256, patternLength - 1)) % PRIME;
    newHash = (newHash * 256 + str[newIndex]) % PRIME;
    return newHash < 0 ? newHash + PRIME : newHash;
}

void rabinKarpSearch(char *text, char *pattern) {
    int textLength = strlen(text);
    int patternLength = strlen(pattern);

    long patternHash = calculateHash(pattern, patternLength);
    long textHash = calculateHash(text, patternLength);

    for (int i = 0; i <= textLength - patternLength; i++) {
        if (patternHash == textHash) {
            int j;
            for (j = 0; j < patternLength; j++) {
                if (text[i + j] != pattern[j]) {
                    break;
                }
            }
            if (j == patternLength) {
                printf("Pattern found at index %d\n", i);
            }
        }

        if (i < textLength - patternLength) {
            textHash = recalculateHash(text, i, i + patternLength, textHash, patternLength);
        }
    }
}

int main() {
    char text[] = "ABABDABACDABABCABAB";
    char pattern[] = "ABABCABAB";

    printf("Text: %s\n", text);
    printf("Pattern: %s\n", pattern);

    rabinKarpSearch(text, pattern);

    return 0;
}
